lucene/core/src/java/org/apache/lucene/index/SegmentInfos.java

```diff
@@ -374,7 +376,15 @@ public static final SegmentInfos readCommit(Directory directory, ChecksumIndexIn
       if (softDelCount + delCount > info.maxDoc()) {
         throw new CorruptIndexException("invalid deletion count: " + softDelCount + delCount + " vs maxDoc=" + info.maxDoc(), input);
       }
-      SegmentCommitInfo siPerCommit = new SegmentCommitInfo(info, delCount, softDelCount, delGen, fieldInfosGen, dvGen);
+      final byte[] sciId;
+      if (format > VERSION_74) {
+        sciId = new byte[StringHelper.ID_LENGTH];
+        input.readBytes(sciId, 0, sciId.length);
+      } else {
+        sciId = infos.id;
+        // NOCOMMIT can we do this? it would at least give us consistent BWC but we can't identify the same SCI in different commits
```

Comment:
Good question ... maybe we could use `info.getId()`?   Then it'd be unique across `SCI`, but, shared across `SI` which is also weird.

Reply:
I don't think we should use info.getId() since that would mean that the same SegmentInfo instances are treated the same even if two IW made changes to it and it's generations. The way we have it now it's only considered the same if the overall commit is the same which is good i guess?

Reply:
When we introduced SegmentInfos#getId, we returned `null` as an ID for older segments. This is probably a safer option here as well, as callers can fall back to whatever behavior makes sense for them such as using a strong hash of the commit files as an ID, or re-downloading all files of the commit all the time and giving up incrementality?

Reply:
I think this case is a bit more complicated due to the changing nature of this ID. For each change (DV, llive docs, fields) we need to move to a new ID. Should we then just accept null and create a new ID once it changes or should we stick with `null` on these segments until they are written first time? Introducing `null` requires quite some changes in how we handle this which we can do, for sure. I still wonder if we can get away with stealing the _parent_ ID and have a smooth upgrade path.

Reply:
I pushed a new commit https://github.com/apache/lucene-solr/pull/1434/commits/f0a72f82bb17bd2582799aa25514ef764e012570 to address this.

Reply:
I don't understand the idea of stealing the parent ID, wouldn't it cause Lucene to consider commits equal when they are not, which would be a much worse problem than considering commits different when they are equal?

Reply:
Ah sorry I got confused because I thought that "parent" was referring to SegmentInfo (no s) instead of SegmentInfos, but I agree that SegmentInfos is not great either.

Reply:
+1 for `null` and have caller deal with the BWC.  Stealing the `id` from another place (either "parent" `SegmentInfos` or the corresponding `SegmentInfo`) seems too dangerous.